### ISSUE TITLE
Synthetic Chinese String Dataset数据集训练速度

### ISSUE THREAD

**Issue by whwme** [Sierkinhane/crnn_chinese_characters_rec#248]
2080ti显卡训练这个数据集   粗略估算了一下时间 要118天  有什么办法可以加速一下

**Comment by jasnei** [comment]
@whwme
这个是在瓶颈出现在读数据上面，可以改变一下读数据的方式，可以快1倍以上的速度，我16G内存，4G显示，Batch_size, 128，一个epoch大概 5000秒，看作者的，200张图像/s，我没有用prefecter技术之前大概是320张图像/s。用了之后750张/s。如果内存够大，你可以把batch_size 增大。

![QQ浏览器截图20201026085903](https://user-images.githubusercontent.com/52853052/97124080-a9a37980-1769-11eb-89d5-6e40c813f79a.png)
修改lib/core/function.py
整个文件都修改了。
代码如下：
`
"""
利用了prefetcher技术，连计算的同时也在读取数据
虽然现在GPU利用率比之前还低，但处理速度，比之前快了一倍
缺点：内存占用比之前多一倍
"""

from  __future__ import  absolute_import
import time
import lib.utils.utils as utils
import torch


class AverageMeter(object):

    """Computes and stores the average and current value"""
    def __init__(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count

class data_prefetcher():

    def __init__(self, loader):
        self.loader = iter(loader)
        self.stream = torch.cuda.Stream()
        # 在数据处理里已经做了标准化了，所以这里不再需要做了
        # self.mean = torch.tensor([0.485 * 255, 0.456 * 255, 0.406 * 255]).cuda().view(1,3,1,1)
        # self.std = torch.tensor([0.229 * 255, 0.224 * 255, 0.225 * 255]).cuda().view(1,3,1,1)

        # self.mean = torch.tensor([0.485 * 255, 0.456 * 255, 0.406 * 255]).cuda().view(1,3,1,1)
        # self.std = torch.tensor([0.229 * 255, 0.224 * 255, 0.225 * 255]).cuda().view(1,3,1,1)
        # With Amp, it isn't necessary to manually convert data to half.
        # if args.fp16:
        #     self.mean = self.mean.half()
        #     self.std = self.std.half()
        self.preload()

    def preload(self):
        try:
            self.next_input, self.next_target = next(self.loader)
        except StopIteration:
            self.next_input = None
            self.next_target = None
            return
        with torch.cuda.stream(self.stream):
            self.next_input = self.next_input.cuda(non_blocking=True)
            self.next_target = self.next_target.cuda(non_blocking=True)
            # With Amp, it isn't necessary to manually convert data to half.
            # if args.fp16:
            #     self.next_input = self.next_input.half()
            # else:
            self.next_input = self.next_input.float()
            # self.next_input = self.next_input.sub_(self.mean).div_(self.std)
            
    def next(self):
        torch.cuda.current_stream().wait_stream(self.stream)
        input = self.next_input
        target = self.next_target
        self.preload()
        return input, target
        
        
def train(config, train_loader, dataset, converter, model, criterion, optimizer, device, epoch, writer_dict=None, output_dict=None):

    batch_time = AverageMeter()
    data_time = AverageMeter()
    losses = AverageMeter()

    model.train()

    end = time.time()
    
    prefetcher = data_prefetcher(train_loader)
    inp, idx = prefetcher.next()

    i = 0
    while inp is not None:
        i += 1
        # measure data time
        data_time.update(time.time() - end)
        
        labels = utils.get_batch_label(dataset, idx)
        inp = inp.to(device)

        # inference
        preds = model(inp)

        # compute loss
        batch_size = inp.size(0)
        # print(batch_size, len(labels))
        text, length = converter.encode(labels)                    # length = 一个batch中的总字符长度, text = 一个batch中的字符所对应的下标
        preds_size = torch.IntTensor([preds.size(0)] * batch_size) # timestep * batchsize
        loss = criterion(preds, text, preds_size, length)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        losses.update(loss.item(), inp.size(0))

        batch_time.update(time.time()-end)
        if i % config.PRINT_FREQ == 0:
            msg = 'Epoch: [{0}][{1}/{2}]\t' \
                  'Time {batch_time.val:.3f}s ({batch_time.avg:.3f}s)\t' \
                  'Speed {speed:.1f} samples/s\t' \
                  'Data {data_time.val:.3f}s ({data_time.avg:.3f}s)\t' \
                  'Loss {loss.val:.5f} ({loss.avg:.5f})\t'.format(
                      epoch, i, len(train_loader), batch_time=batch_time,
                      speed=inp.size(0)/batch_time.val,
                      data_time=data_time, loss=losses)
            print(msg)

            if writer_dict:
                writer = writer_dict['writer']
                global_steps = writer_dict['train_global_steps']
                writer.add_scalar('train_loss', losses.avg, global_steps)
                writer_dict['train_global_steps'] = global_steps + 1

        end = time.time()
        
        inp, idx = prefetcher.next()


def validate(config, val_loader, dataset, converter, model, criterion, device, epoch, writer_dict, output_dict):

    losses = AverageMeter()
    model.eval()

    n_correct = 0
    with torch.no_grad():
        for i, (inp, idx) in enumerate(val_loader):

            labels = utils.get_batch_label(dataset, idx)
            inp = inp.to(device)

            # inference
            preds = model(inp)

            # compute loss
            batch_size = inp.size(0)
            text, length = converter.encode(labels)
            preds_size = torch.IntTensor([preds.size(0)] * batch_size)
            loss = criterion(preds, text, preds_size, length)

            losses.update(loss.item(), inp.size(0))

            _, preds = preds.max(2)
            preds = preds.transpose(1, 0).contiguous().view(-1)
            sim_preds = converter.decode(preds.data, preds_size.data, raw=False)
            for pred, target in zip(sim_preds, labels):
                if pred == target:
                    n_correct += 1

            if (i + 1) % config.PRINT_FREQ == 0:
                print('Epoch: [{0}][{1}/{2}]'.format(epoch, i, len(val_loader)))

            if i == config.TEST.NUM_TEST_BATCH:
                break

    raw_preds = converter.decode(preds.data, preds_size.data, raw=True)[:config.TEST.NUM_TEST_DISP]
    for raw_pred, pred, gt in zip(raw_preds, sim_preds, labels):
        print('%-20s => %-20s, gt: %-20s' % (raw_pred, pred, gt))

    num_test_sample = config.TEST.NUM_TEST_BATCH * config.TEST.BATCH_SIZE_PER_GPU
    if num_test_sample > len(dataset):
        num_test_sample = len(dataset)

    print("[#correct:{} / #total:{}]".format(n_correct, num_test_sample))
    accuracy = n_correct / float(num_test_sample)
    print('Test loss: {:.4f}, accuray: {:.4f}'.format(losses.avg, accuracy))

    if writer_dict:
        writer = writer_dict['writer']
        global_steps = writer_dict['valid_global_steps']
        writer.add_scalar('valid_acc', accuracy, global_steps)
        writer_dict['valid_global_steps'] = global_steps + 1

    return accuracy
`

**Comment by Donaghys** [comment]
> @whwme
> 这个是在瓶颈出现在读数据上面，可以改变一下读数据的方式，可以快1倍以上的速度，我16G内存，4G显示，Batch_size, 128，一个epoch大概 5000秒，看作者的，200张图像/s，我没有用prefecter技术之前大概是320张图像/s。用了之后750张/s。如果内存够大，你可以把batch_size 增大。
> 
> ![QQ浏览器截图20201026085903](https://user-images.githubusercontent.com/52853052/97124080-a9a37980-1769-11eb-89d5-6e40c813f79a.png)
> 修改lib/core/function.py
> 整个文件都修改了。
> 代码如下：
> `
> """
> 利用了prefetcher技术，连计算的同时也在读取数据
> 虽然现在GPU利用率比之前还低，但处理速度，比之前快了一倍
> 缺点：内存占用比之前多一倍
> """
> 
> from **future** import absolute_import
> import time
> import lib.utils.utils as utils
> import torch
> 
> class AverageMeter(object):
> 
> ```
> """Computes and stores the average and current value"""
> def __init__(self):
>     self.val = 0
>     self.avg = 0
>     self.sum = 0
>     self.count = 0
>     self.reset()
> 
> def reset(self):
>     self.val = 0
>     self.avg = 0
>     self.sum = 0
>     self.count = 0
> 
> def update(self, val, n=1):
>     self.val = val
>     self.sum += val * n
>     self.count += n
>     self.avg = self.sum / self.count
> ```
> 
> class data_prefetcher():
> 
> ```
> def __init__(self, loader):
>     self.loader = iter(loader)
>     self.stream = torch.cuda.Stream()
>     # 在数据处理里已经做了标准化了，所以这里不再需要做了
>     # self.mean = torch.tensor([0.485 * 255, 0.456 * 255, 0.406 * 255]).cuda().view(1,3,1,1)
>     # self.std = torch.tensor([0.229 * 255, 0.224 * 255, 0.225 * 255]).cuda().view(1,3,1,1)
> 
>     # self.mean = torch.tensor([0.485 * 255, 0.456 * 255, 0.406 * 255]).cuda().view(1,3,1,1)
>     # self.std = torch.tensor([0.229 * 255, 0.224 * 255, 0.225 * 255]).cuda().view(1,3,1,1)
>     # With Amp, it isn't necessary to manually convert data to half.
>     # if args.fp16:
>     #     self.mean = self.mean.half()
>     #     self.std = self.std.half()
>     self.preload()
> 
> def preload(self):
>     try:
>         self.next_input, self.next_target = next(self.loader)
>     except StopIteration:
>         self.next_input = None
>         self.next_target = None
>         return
>     with torch.cuda.stream(self.stream):
>         self.next_input = self.next_input.cuda(non_blocking=True)
>         self.next_target = self.next_target.cuda(non_blocking=True)
>         # With Amp, it isn't necessary to manually convert data to half.
>         # if args.fp16:
>         #     self.next_input = self.next_input.half()
>         # else:
>         self.next_input = self.next_input.float()
>         # self.next_input = self.next_input.sub_(self.mean).div_(self.std)
>         
> def next(self):
>     torch.cuda.current_stream().wait_stream(self.stream)
>     input = self.next_input
>     target = self.next_target
>     self.preload()
>     return input, target
> ```
> 
> def train(config, train_loader, dataset, converter, model, criterion, optimizer, device, epoch, writer_dict=None, output_dict=None):
> 
> ```
> batch_time = AverageMeter()
> data_time = AverageMeter()
> losses = AverageMeter()
> 
> model.train()
> 
> end = time.time()
> 
> prefetcher = data_prefetcher(train_loader)
> inp, idx = prefetcher.next()
> 
> i = 0
> while inp is not None:
>     i += 1
>     # measure data time
>     data_time.update(time.time() - end)
>     
>     labels = utils.get_batch_label(dataset, idx)
>     inp = inp.to(device)
> 
>     # inference
>     preds = model(inp)
> 
>     # compute loss
>     batch_size = inp.size(0)
>     # print(batch_size, len(labels))
>     text, length = converter.encode(labels)                    # length = 一个batch中的总字符长度, text = 一个batch中的字符所对应的下标
>     preds_size = torch.IntTensor([preds.size(0)] * batch_size) # timestep * batchsize
>     loss = criterion(preds, text, preds_size, length)
> 
>     optimizer.zero_grad()
>     loss.backward()
>     optimizer.step()
> 
>     losses.update(loss.item(), inp.size(0))
> 
>     batch_time.update(time.time()-end)
>     if i % config.PRINT_FREQ == 0:
>         msg = 'Epoch: [{0}][{1}/{2}]\t' \
>               'Time {batch_time.val:.3f}s ({batch_time.avg:.3f}s)\t' \
>               'Speed {speed:.1f} samples/s\t' \
>               'Data {data_time.val:.3f}s ({data_time.avg:.3f}s)\t' \
>               'Loss {loss.val:.5f} ({loss.avg:.5f})\t'.format(
>                   epoch, i, len(train_loader), batch_time=batch_time,
>                   speed=inp.size(0)/batch_time.val,
>                   data_time=data_time, loss=losses)
>         print(msg)
> 
>         if writer_dict:
>             writer = writer_dict['writer']
>             global_steps = writer_dict['train_global_steps']
>             writer.add_scalar('train_loss', losses.avg, global_steps)
>             writer_dict['train_global_steps'] = global_steps + 1
> 
>     end = time.time()
>     
>     inp, idx = prefetcher.next()
> ```
> 
> def validate(config, val_loader, dataset, converter, model, criterion, device, epoch, writer_dict, output_dict):
> 
> ```
> losses = AverageMeter()
> model.eval()
> 
> n_correct = 0
> with torch.no_grad():
>     for i, (inp, idx) in enumerate(val_loader):
> 
>         labels = utils.get_batch_label(dataset, idx)
>         inp = inp.to(device)
> 
>         # inference
>         preds = model(inp)
> 
>         # compute loss
>         batch_size = inp.size(0)
>         text, length = converter.encode(labels)
>         preds_size = torch.IntTensor([preds.size(0)] * batch_size)
>         loss = criterion(preds, text, preds_size, length)
> 
>         losses.update(loss.item(), inp.size(0))
> 
>         _, preds = preds.max(2)
>         preds = preds.transpose(1, 0).contiguous().view(-1)
>         sim_preds = converter.decode(preds.data, preds_size.data, raw=False)
>         for pred, target in zip(sim_preds, labels):
>             if pred == target:
>                 n_correct += 1
> 
>         if (i + 1) % config.PRINT_FREQ == 0:
>             print('Epoch: [{0}][{1}/{2}]'.format(epoch, i, len(val_loader)))
> 
>         if i == config.TEST.NUM_TEST_BATCH:
>             break
> 
> raw_preds = converter.decode(preds.data, preds_size.data, raw=True)[:config.TEST.NUM_TEST_DISP]
> for raw_pred, pred, gt in zip(raw_preds, sim_preds, labels):
>     print('%-20s => %-20s, gt: %-20s' % (raw_pred, pred, gt))
> 
> num_test_sample = config.TEST.NUM_TEST_BATCH * config.TEST.BATCH_SIZE_PER_GPU
> if num_test_sample > len(dataset):
>     num_test_sample = len(dataset)
> 
> print("[#correct:{} / #total:{}]".format(n_correct, num_test_sample))
> accuracy = n_correct / float(num_test_sample)
> print('Test loss: {:.4f}, accuray: {:.4f}'.format(losses.avg, accuracy))
> 
> if writer_dict:
>     writer = writer_dict['writer']
>     global_steps = writer_dict['valid_global_steps']
>     writer.add_scalar('valid_acc', accuracy, global_steps)
>     writer_dict['valid_global_steps'] = global_steps + 1
> 
> return accuracy
> ```
> 
> `

您好，请问这样子会不会影响最终模型的效果？

**Comment by Donaghys** [comment]
![image](https://user-images.githubusercontent.com/49194956/99139807-ee3e7880-2676-11eb-9d81-1ee1e5c83011.png)
速度是快了不少，但是loss降不下去，平时最终在0.0005收敛，现在在0.6就收敛了，有好的解决办法吗？

**Comment by jasnei** [comment]
这个不会影响模型的，这只是dataloader会快一点。你可以测试一下准确率看看效果。这个跟原来的没有什么区别，区别是是加了prefecter技术，读数据时候快点。应该不影响到loss还有准确率的。你可以对比一下两个模型的效果。

**Comment by Donaghys** [comment]
> 这个不会影响模型的，这只是dataloader会快一点。你可以测试一下准确率看看效果。
感谢您的工作，同样使用作者提供的预训练模型finetune，如果我不使用prefecter，我的数据一个bs后就可以到2.多，只需要2-3个epoch就可以下降到很低，大约在0.005左右，验证集有90%以上的准确率，但是速度在500/s张左右，加入这个prefecter后，速度到了3100/s，但是这个损失就下降的很慢，现在已经30个epoch了，损失依然是0.6，曲线如刚才发的样子，我现在怀疑网络已经收敛了~，但是除了这个dataloader以外，我没有对代码做任何修改，我很困惑，希望得到您的指点~

**Comment by Donaghys** [comment]
> 这个不会影响模型的，这只是dataloader会快一点。你可以测试一下准确率看看效果。这个跟原来的没有什么区别，区别是是加了prefecter技术，读数据时候快点。应该不影响到loss还有准确率的。你可以对比一下两个模型的效果。
主要还是mode(inp),以前是cpu做预测，您这是gpu做预测~，两者会不会有差距呢？
